llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1499,7 +1499,7 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
           return replaceInstUsesWith(I, X);
 
         // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
-        if (Op0->hasOneUse()) {
+        if (Op0->hasOneUse() && isGuaranteedNotToBeUndef(X, &AC, &I, &DT)) {
```

Comment:
We should do `if (!isGuaranteedNotToBeUndef) CreateFreeze` instead.

Reply:
> We should do `if (!isGuaranteedNotToBeUndef) CreateFreeze` instead.

That creates an extra instruction, so we end up with more instructions than we started with.

Reply:
It's ok if the extra instruction is freeze.